File: src/js/minimax-player.js
class MinimaxBoardNode {
    constructor(boardState, playerIndex) {
        this.boardState = boardState;
        this.score = 0;
        this.playerIndex = playerIndex;
        this.oppIndex = 1 - playerIndex;
        let pos1 = boardState.playerPositions[0];
        let pos2 = boardState.playerPositions[1];
        this.playerPos = playerIndex == 0 ? pos1 : pos2;
        this.oppPos = playerIndex == 0 ? pos2 : pos1;
        this.playerWalls = boardState.playerWallCounts[playerIndex];
        this.action = null;
    }

    buildChildren(depth, scoringPlayer, maximizing, alpha, beta) {
        let oppIndex = 1 - scoringPlayer;
        let oppDist = this.boardState.getPlayerDistance(oppIndex);
        let playerDist = this.boardState.getPlayerDistance(scoringPlayer);
        if (playerDist == 0 || oppDist == 0 || depth == 0) {
            // When the board has no children calculate the distances from the end for each player.
            this.score = oppDist - playerDist;
            return;
        }

        // Add all the valid movement positions to the valid moves.
        let validActions = Array.from(this.boardState.getValidMoves(this.playerPos, this.oppPos), x => new MoveAction(x));

        if (this.playerWalls > 0) {
            // For each column
            for (let x = 0; x < 8; x++) {
                // For each row
                for (let y = 0; y < 8; y++) {
                    let pos = new Vector(x, y);
                    // For each orientation
                    for (let o = 1; o <= 2; o++) {
                        // If this is a valid place to put a wall.
                        if (isValidWall(this.boardState, pos, o)) {
                            validActions.push(new BlockAction(pos, o));
                        }
                    }
                }
            }
        }

        let a = alpha;
        let b = beta;
        let value = maximizing ? -Infinity : Infinity;
        for (let action of validActions) {
            let newBoardState = this.boardState.copy();
            action.apply(newBoardState, this.playerIndex);
            if (!isEitherPlayerTrapped(newBoardState)) {
                let childNode = new MinimaxBoardNode(newBoardState, this.oppIndex);
                childNode.buildChildren(depth - 1, scoringPlayer, !maximizing, a, b);
                if (maximizing) {
                    if (childNode.score > value) {
                        value = childNode.score;
                        this.action = action;
                    }
                    a = Math.max(a, value);
                    if (a >= b)
                        break;
                }
                else {
                    if (childNode.score < value) {
                        value = childNode.score;
                        this.action = action;
                    }
                    b = Math.min(b, value);
                    if (a >= b)
                        break;
                }
            }
        }
        this.score = value;
    }
}

class MinimaxPlayer extends Player {

    constructor(depth) {
        super();
        this.branchDepth = depth;
    }

    takeAction(boardState) {
        let currentNode = new MinimaxBoardNode(boardState, this.index);
        currentNode.buildChildren(this.branchDepth, this.index, true, -Infinity, Infinity);
        return currentNode.action;
    }
}

File: src/rs/minimax_player.rs
use crate::action::Action;
use crate::board_state::BoardState;
use crate::validation::*;

use std::f32;

pub struct MinimaxPlayer {}

impl MinimaxPlayer {
    pub fn take_action(board_state: &BoardState, player_index: usize, branch_depth: usize) -> Action {
        let mut node = MinimaxBoardNode::new(board_state, player_index);
        let _turn_play_count = node.build_children(branch_depth, player_index, true, f32::MIN, f32::MAX);
        return node.best_action.unwrap();
    }
}

struct MinimaxBoardNode<'a> {
    pub board_state: &'a BoardState,
    pub player_index: usize,
    pub best_action: Option<Action>,
    pub score: f32,
}

impl<'a> MinimaxBoardNode<'a> {
    pub fn new(board_state: &'a BoardState, player_index: usize) -> Self {
        MinimaxBoardNode {
            board_state: board_state,
            player_index: player_index,
            best_action: None,
            score: 0.0,
        }
    }

    pub fn build_children(
        &mut self,
        branch_depth: usize,
        scoring_player: usize,
        maximizing: bool,
        alpha: f32,
        beta: f32,
    ) -> usize {
        let opp_index = 1 - scoring_player;
        let opp_distance = self.board_state.get_player_distance(opp_index);
        let player_distance = self.board_state.get_player_distance(scoring_player);
        if player_distance == 0 || opp_distance == 0 || branch_depth == 0 {
            // When the board has no children calculate the distances from the end for each player.
            self.score = (opp_distance - player_distance) as f32;
            return 1;
        }

        let mut valid_actions = Vec::<Action>::new();
        valid_actions.append(&mut get_valid_move_actions(&self.board_state, self.player_index));
        valid_actions.append(&mut get_valid_block_actions(&self.board_state, self.player_index));

        let mut a = alpha;
        let mut b = beta;
        let mut score = if maximizing { f32::MIN } else { f32::MAX };
        let mut best_action_index = 0;
        let mut turn_play_count = 1;
        for (i, &action) in valid_actions.iter().enumerate() {
            let mut new_board_state = self.board_state.clone();
            action.apply(&mut new_board_state, self.player_index);
            if !is_either_player_trapped(&new_board_state) {
                let mut child_node = MinimaxBoardNode::new(&new_board_state, 1 - self.player_index);
                turn_play_count += child_node.build_children(branch_depth - 1, scoring_player, !maximizing, a, b);
                if maximizing {
                    if child_node.score > score {
                        score = child_node.score;
                        best_action_index = i;
                    }
                    a = f32::max(a, score);
                    if a >= b {
                        break;
                    }
                } else {
                    if child_node.score < score {
                        score = child_node.score;
                        best_action_index = i;
                    }
                    b = f32::min(b, score);
                    if a >= b {
                        break;
                    }
                }
            }
        }

        self.best_action = Option::from(valid_actions[best_action_index]);
        self.score = score;
        return turn_play_count;
    }
}
